File: 63-unique-paths-ii/63-unique-paths-ii.java
class Solution {
    public int uniquePathsWithObstacles(int[][] grid) {
        int m = grid.length, n = grid[0].length;
        int[][] dp = new int[m + 1][n + 1];
        if(grid[0][0] == 1 || grid[m - 1][n - 1] == 1) return 0;
                
        for(int i = 1; i <= m; i++) {
            for(int j = 1; j <= n; j++) {
                if(i == 1 && j == 1) {
                    dp[i][j] = 1;
                }
                else if(grid[i-1][j-1] != 1) {
                    dp[i][j] = dp[i - 1][j] + dp[i][j - 1];
                } else dp[i][j] = 0;
            }
        }
        return dp[m][n];
    }
}

File: 63-unique-paths-ii/63-unique-paths-ii.py
class Solution:
    def uniquePathsWithObstacles(self, grid: List[List[int]]) -> int:
        m, n = len(grid), len(grid[0])
        dp = [[0] * n for _ in range(m)]
        if grid[0][0] == 1: return 0
        
        dp[0][0] = 1
        for i in range(m):
            for j in range(n):
                if grid[i][j] != 1:
                    if i == j == 0: continue
                    elif i == 0 or j == 0:
                        if not i: dp[i][j] = dp[i][j - 1]
                        if not j: dp[i][j] = dp[i - 1][j]
                    else:
                        dp[i][j] += dp[i- 1][j] + dp[i][j - 1]
        return dp[-1][-1]
                    